### PR TITLE
[flutter_tools]  surgically remove outputs from shared directory

### LINKED ISSUE #53667: Write a file containing the last build configuration into the output directory
See https://github.com/flutter/flutter/issues/53546
See https://github.com/flutter/flutter/pull/53601#issuecomment-606737687

As a shorter term solution to multiple builds overlapping in output directories, we can perform a sanity check by writing the last config hash into the output root. If it is different, blow away the folder before writing. Also need to check that nothing else is putting files there.

### LINKED ISSUE #53668: Remove shared output directories from build system
See https://github.com/flutter/flutter/issues/53546
See https://github.com/flutter/flutter/pull/53601#issuecomment-606737687

As a longer term solution, only use the unique build directories - removing the need for https://github.com/flutter/flutter/issues/53667

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

Fixes #53668
Fixes #53667

When multiple build configurations output to the same directory, the per-target output removal does not work properly, because they do not have knowledge of what the other configuration output.

To solve this, write a file into the output directory pointing to the last configuration. If this exists and is different from the current configuration, then parse that configurations outputs and surgically remove any that are not in the current output set

### CODE REVIEW

File: packages/flutter_tools/lib/src/build_system/targets/macos.dart [hunk 1]
```diff
@@ -56,9 +56,8 @@ abstract class UnpackMacOS extends Target {
     final Directory targetDirectory = environment
       .outputDir
       .childDirectory('FlutterMacOS.framework');
-    // This is necessary because multiple different build configurations will
-    // output different files here. Build cleaning only works when the files
-    // change within a build configuration.
+    // Deleting this directory is required or else the FlutterMacOS module
+    // cannot be found.
     if (targetDirectory.existsSync()) {
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/build_system/targets/macos.dart:61]: Deleting this directory is still required, though I'm not sure why yet

**Comment by jonahwilliams** [comment]
Same as previous PR except I reverted the change to delete the engine output directory for macOS

File: packages/flutter_tools/lib/src/build_system/targets/macos.dart (hunk 1 above)
Comment (jmagman) [reply packages/flutter_tools/lib/src/build_system/targets/macos.dart:61]: 🤔

**Review by jmagman — APPROVED** [review]
LGTM

**Comment by jonahwilliams** [comment]
Build is actually green, merging

**Pull request merged**